aws/asg.go

```diff
@@ -213,67 +206,56 @@ func updateTargetGroupsForAutoScalingGroup(svc autoscalingiface.AutoScalingAPI,
 		}
 
 		if len(detachARNs) > 0 {
-			err = detachTargetGroupsFromAutoScalingGroup(svc, detachARNs, autoScalingGroupName)
+			err := detachTargetGroupsFromAutoScalingGroup(svc, detachARNs, autoScalingGroupName)
 			if err != nil {
 				return err
 			}
 		}
 	}
 
-	// limit target group updates to 10 at a time since this is the limit
-	// in AWS API.
-	chunkSize := 10
-
-	for i := 0; i < len(targetGroupARNs); i += chunkSize {
-		end := i + chunkSize
-
-		if end > len(targetGroupARNs) {
-			end = len(targetGroupARNs)
+	attachARNs := make([]string, 0, len(targetGroupARNs))
+	for _, tgARN := range targetGroupARNs {
+		if _, ok := allTGs[tgARN]; ok {
+			attachARNs = append(attachARNs, tgARN)
+		} else {
+			// TODO: it is better to validate stack's target groups earlier to identify owning stack
+			log.Errorf("target group %q does not exist, will not attach", tgARN)
 		}
+	}
 
-		groups := targetGroupARNs[i:end]
-		if len(groups) > 0 {
-			attachParams := &autoscaling.AttachLoadBalancerTargetGroupsInput{
-				AutoScalingGroupName: aws.String(autoScalingGroupName),
-				TargetGroupARNs:      aws.StringSlice(groups),
-			}
-			_, err = svc.AttachLoadBalancerTargetGroups(attachParams)
-			if err != nil {
-				return err
-			}
+	if len(attachARNs) > 0 {
+		err := attachTargetGroupsToAutoScalingGroup(svc, attachARNs, autoScalingGroupName)
+		if err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
-func describeTagsChunked(svc elbv2iface.ELBV2API, arns []string) ([]*elbv2.TagDescription, error) {
+func describeTags(svc elbv2iface.ELBV2API, arns []string) ([]*elbv2.TagDescription, error) {
 	descs := make([]*elbv2.TagDescription, 0, len(arns))
-	chunkSize := 20
-
-	for i := 0; i < len(arns); i += chunkSize {
-		end := i + chunkSize
-
-		if end > len(arns) {
-			end = len(arns)
-		}
-
-		arnsChunked := arns[i:end]
-		if len(arnsChunked) > 0 {
-			tgParams := &elbv2.DescribeTagsInput{
-				ResourceArns: aws.StringSlice(arnsChunked),
-			}
-
-			tgResp, err := svc.DescribeTags(tgParams)
-			if err != nil {
-				return nil, err
-			}
-
+	err := processChunked(arns, 20, func(chunk []string) error {
```

Comment:
I wonder why it's 20 here :confused:

Reply:
I have added a link to the docs that specify the limit